File: problems/897-increasing-order-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode head = new TreeNode();
    TreeNode tail = head;
    public TreeNode increasingBST(TreeNode root) {
        inorder(root);
        return head.right;
    }
    public void inorder(TreeNode node){
        if(node == null) return;
        inorder(node.left);
        tail.right = new TreeNode(node.val);
        tail = tail.right;
        inorder(node.right);
    }
}

File: problems/897-increasing-order-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        TreeNode* res = new TreeNode(0);
        TreeNode* cur = res;
        function<void(TreeNode*)> dfs = [&](TreeNode* node) {
            if (!node) return;
            dfs(node->left);
            cur->right = new TreeNode(node->val);
            cur = cur->right;
            dfs(node->right);
        };
        dfs(root);
        return res->right;
    }
};